### ISSUE TITLE
Completion UI considerations

### ISSUE THREAD

**Issue by SebastianSzturo** [gobijan/nova-copilot-lsp#3]
Thanks for starting this project, really appreciate someone trying to port the LSP plugin from vim. (I did try a year ago and gave up! :P)

One thing I ran into while exploring this last year, is that Nova doesn't really have the same UI concepts as VSCode. The reason why VSCode's completion are so seamless are due to them being rendered as translucent comments that is easily readable and dismissable:

<img width="1021" alt="CleanShot 2023-04-28 at 13 32 43@2x" src="https://user-images.githubusercontent.com/2699745/235055051-638237b5-9c03-4d03-8f51-81955ea40650.png">

For Nova we really only have the [Language Completion](https://docs.nova.app/syntax-reference/completions/) concept that might not be ideal for longer codeblocks:

<img width="736" alt="CleanShot 2023-04-28 at 13 35 38@2x" src="https://user-images.githubusercontent.com/2699745/235055151-4bb3d08c-0e64-4a71-a47b-9f2e4894c9e4.png">

Have you put any thought into this? I guess for now Language Completions is better than nothing and maybe we can get the attention from Panic to do something about this.

**Comment by gobijan** [comment]
Actually a good point. I thought that maybe something in the sidebar could work? What you think could work best?

**Comment by kitsah** [comment]
How about placing suggestions in both a Sidebar and the Language Completion drop-down? That way, we can see a short preview of the suggested code below our cursor as we type, then we can check for the full code block in the sidebar.

This way, we can still accept suggestions simply by using Tab, as we are used to, and we can still see the full code suggestion in a Sidebar.

A potential issue is whether or not the Sidebar supports syntax highlighting?

**Comment by SkelegonDK** [comment]
It has to be like in VS code. But as a temporary solution I would put "Copilot" as the top completion option and show the code in the sidebar. That said, having a sidebar open ALL the time takes a lot of valuable horizontal space. However, in VS Code you have to explicitly tell VS code to generate suggestions based on a comment, and it will give you up to 10 options. 
What if any comment starting with (python) `#COPILOT: Your prompt definition` Automatically outputs in the sidebar. That way if you are the kind of person that only wants occasional suggestions, you can opt for that alone.

**Comment by ctkjose** [comment]
Extensions are quite limited. There are currently no UI options for plugins. A sidebar is limited to a tree view which is not any better than the auto completion menu. You can use a NotificationRequest to prompt the user for a query and populate the sidebar. I honestly don't see an elegant UX and interaction.

I posted a feature [request](https://devforum.nova.app/t/what-i-wish-nova-provided/1687) in the forum for the ability to build UI.

**Comment by SebastianSzturo** [comment]
I've also created a request to see if we can get a ["ghost text" ](https://code.visualstudio.com/blogs/2023/03/30/vscode-copilot)completion interface like VSCode: https://devforum.nova.app/t/ghost-text-for-copilot-completions/2195

**Comment by SebastianSzturo** [comment]
@wptechnology PRs are welcome ;)

**Comment by net-tech** [comment]
> That said, having a sidebar open ALL the time takes a lot of valuable horizontal space

Couldn't you just have multiple sidebars open? At the end of the day even if it takes a lot of space it is nicer than the current way it works.

**Comment by net-tech** [comment]
> > That said, having a sidebar open ALL the time takes a lot of valuable horizontal space
> 
> Couldn't you just have multiple sidebars open? At the end of the day even if it takes a lot of space it is nicer than the current way it works.

@SebastianSzturo Any thoughts on this suggestion? Maybe it could be a config option?

**Comment by gobijan** [comment]
Yeah I agree. Not willing to bet on a closed source product that is semi maintained.

**Comment by realCrush** [comment]
Are there any progress on this issue? Is it possible to have a similar experience to Copilot in VSCode by 2024? Honestly, I don't understand why AI auto-complete, a very important and basic function today, is still not being supported. It really affects user experience.

**Comment by andreiborisov** [comment]
Beta for Nova 12 is released, is there a suitable API added? https://nova.app/releases/
